### DISCUSSION TITLE
Considering an alternative to inlinestyler

### DISCUSSION THREAD

**Discussion by Stranger6667** [pretalx/pretalx#1256] [Ideas] [answered]
Hello!

I noticed that `pretalx` uses [inlinestyler](https://github.com/dlanger/inlinestyler) where [the last commit](https://github.com/dlanger/inlinestyler/commit/335c4fbab892f0ed67466a6beaea6a91f395ad12) was in Mar 2018, so `inlinestyler` unfortunately looks abandoned (I was using it myself for quite some time). Therefore I'd like to propose replacing it with [css-inline](https://github.com/Stranger6667/css-inline/tree/master/bindings/python) (that I authored) or some other maintained alternative (e.g. [premailer](https://github.com/peterbe/premailer)). 

`css-inline` is significantly faster - a sample HTML constructed from rendered `mailwrapper.html` is processed locally in ~130µs vs. ~20ms with `inlinestyler` vs. ~3ms with `premailer`. Though, it adds a dependency on the Rust toolchain for some users (where wheels aren't provided, like arm64) and generally supports fewer architectures than `lxml`-based libraries (`inlineslyler` & `premailer`). However, I don't know whether it is a concern. The replacement itself would look like this:

```python
from css_inline import inline

inline(html, remove_style_tags=True)
```

Let me know what do you think :) If you'll be ok with replacing `inlinestyler`, I'd be happy to submit a PR

**Comment by rixx** [comment, marked as answer]
Hm, that's interesting. I see you're building macos and manylinux wheels - do you know if those work for BSD systems? I know we have several users who run pretalx on BSD, and I'm hesitant to add Rust to their install requirements.

As long as the wheels are available (and will be, in the future) for common platforms, I don't see a problem, and would welcome a pull request.

**Reply by Stranger6667** [reply]
Thank you for providing context!

> As long as the wheels are available (and will be, in the future) for common platforms, I don't see a problem, and would welcome a pull request.

FYI, I updated the build matrix a bit to support MUSL Linux, macOS ARM64, and some more platforms.

> do you know if those work for BSD systems?

Unfortunately, [the official docs](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#platform-tags-for-other-nix-platforms) do not say much about wheels for *nix platforms other than Linux/Mac/AIX, and I am inclined to think that supporting BSD platforms would require some substantial amount of work on multiple layers (though e.g. FreeBSD 64bit is supported by rustc) :( I will check if it is possible in the PyO3 (`css-inline` uses it to provide Python bindings) support channel.

I think, that adding a dependency on the Rust toolchain for BSD users might lead to unnecessary complications in their installation process, so I am hesitant as well :) Maybe `premailer` will be a better alternative given the circumstances?

**Reply by Stranger6667** [reply]
Update - it is possible to build `css-inline` on FreeBSD and create a wheel from it, but PyPI won't accept it
